cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewRunnerService.java

```diff
@@ -117,6 +126,9 @@ protected void run() {
   @Override
   protected void shutDown() throws Exception {
     LOG.debug("Preview runner service completed");
+    if (completionNotifier != null) {
+      completionNotifier.call();
```

Comment:
What is this for?? Since this class is a Service, if the caller is interested in the completion of this Service, it should just add a Service Listener to this instance.

Reply:
added listener.